pkg/kube/controllers/extendedsecret/certificatesigningrequest_reconciler.go

```diff
@@ -0,0 +1,209 @@
+	scheme       *runtime.Scheme
+	setReference setReferenceFunc
+}
+
+// Reconcile approves pending CSR and creates its certificate secret
+func (r *ReconcileCertificateSigningRequest) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	instance := &certv1.CertificateSigningRequest{}
+
+	// Set the ctx to be Background, as the top-level context for incoming requests.
+	ctx, cancel := context.WithTimeout(r.ctx, r.config.CtxTimeOut)
+	defer cancel()
+
+	ctxlog.Infof(ctx, "Reconciling certificateSigningRequest '%s'", request.Name)
+	err := r.client.Get(ctx, request.NamespacedName, instance)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			ctxlog.Info(ctx, "Skip reconcile: certificateSigningRequest not found")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		ctxlog.Info(ctx, "Error reading the object")
+		return reconcile.Result{}, errors.Wrap(err, "Error reading certificateSigningRequest")
+	}
+
+	if len(instance.Status.Certificate) != 0 {
+		annotations := instance.GetAnnotations()
+		if annotations == nil {
+			ctxlog.WithEvent(instance, "NotFoundError").Errorf(ctx, "Empty annotations of certificateSigningRequest '%s'", instance.Name)
+			return reconcile.Result{}, nil
+		}
+		secretName, ok := annotations[esv1.AnnotationCertSecretName]
+		if !ok {
+			ctxlog.WithEvent(instance, "NotFoundError").Errorf(ctx, "failed to lookup cert secret name from certificateSigningRequest '%s' annotations", instance.Name)
+			return reconcile.Result{}, nil
+		}
+		namespace, ok := annotations[esv1.AnnotationExSecretNamespace]
+		if !ok {
+			ctxlog.WithEvent(instance, "NotFoundError").Errorf(ctx, "failed to lookup exSecret namespace from certificateSigningRequest '%s' annotations", instance.Name)
+			return reconcile.Result{}, nil
+		}
+
+		privatekeySecret, err := r.getSecret(ctx, namespace, names.PrivateKeySecretName(instance.Name))
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		certSecret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            secretName,
+				Namespace:       namespace,
+				OwnerReferences: instance.OwnerReferences,
+				Labels: map[string]string{
+					esv1.LabelKind: esv1.GeneratedSecretKind,
+				},
+			},
+			Data: map[string][]byte{
+				"certificate": instance.Status.Certificate,
+				"private_key": privatekeySecret.Data["private_key"],
+				"is_ca":       privatekeySecret.Data["is_ca"],
+			},
+		}
+
+		err = r.createSecret(ctx, certSecret)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		err = r.deleteSecret(ctx, privatekeySecret)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		return reconcile.Result{}, nil
+	}
+
+	err = r.approveRequest(ctx, instance.Name)
```

Comment:
I think this should better go into an `else` branch.

Reply:
And couldn't we pass the object directly instead of its name to get rid of another get request?

Reply:
Oh, it's helpful for reducing update conflicts. 
And `r.client` could not update subresource `approval`. we have to use `client-go.client` to update CSR approval directly: 
https://github.com/kubernetes-sigs/controller-runtime/issues/172